e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersTestBaseIT.java

```diff
@@ -0,0 +1,130 @@
+package no.sysco.testing.kafka.e2e;
+
+import org.junit.BeforeClass;
+import org.testcontainers.containers.*;
+import org.testcontainers.containers.wait.strategy.Wait;
+import static org.junit.Assert.fail;
+
+/**
+ * Base test  abstract class contains setup to
+ * initialize testing environment.
+ */
+public abstract class ContainersTestBaseIT {
+
+    static final String TOPIC = "events-message-v1";
+    static final String CONFLUENT_PLATFORM_VERSION = "5.1.1";
+
+    // kafka env
+    static KafkaContainer kafka;
+    static String KAFKA_BROKER_INSIDE_DOCKER_ENV;
+    static GenericContainer schemaRegistry;
+    static String SCHEMA_REGISTRY_INSIDE_DOCKER_ENV;
+
+    // app env
+    static GenericContainer jsonServer;
+    static String JSON_SERVER_INSIDE_DOCKER_ENV;
+    static String JSON_SERVER_EXPOSED;
+
+    static GenericContainer httpProducer;
+    static String HTTP_PRODUCER_EXPOSED;
+
+    static GenericContainer httpMaterializer;
+
+    @BeforeClass
+    public static void start() {
```

Comment:
Since you extracted the base class, be aware that `@BeforeClass` will be triggered for **each** class that extends from this one. In other words, if you have 3 classes `extends ContainersTestBaseIT`, then you will create 3(!!!) networks, Kafka containers, and everything else. And, since you're not using `@AfterClass`, they will stay until the JVM exits.

Consider checking that `start` was already triggered, or use `static {}` block (aka "Singleton containers pattern": https://www.testcontainers.org/test_framework_integration/manual_lifecycle_control/#singleton-containers )

Reply:
That is very useful, thx.